### ISSUE TITLE
Unimplemented method when altering database

### ISSUE THREAD

**Issue by jeffbromberger** [SapientGuardian/SapientGuardian.EntityFrameworkCore.MySql#29]
So far in my development I have been starting clean each time I change my model.  I wipe the database and then run:

add-migration mig1; update-database

This works fine.

But today I went to add a couple of additional fields to one model and then wanted to update the existing database instead of starting from scratch.  The add-migration command was successful, but the subsequent update-database failed with this:

> System.NotImplementedException: The method or operation is not implemented.    at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.Generate(AlterColumnOperation operation, IModel model, MigrationCommandListBuilder builder)
    at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.Generate(MigrationOperation operation, IModel model, MigrationCommandListBuilder builder) 
    at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.Generate(IReadOnlyList`1 operations, IModel model) 
    at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.GenerateUpSql(Migration migration) 
    at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.Migrate(String targetMigration) 
    at Microsoft.EntityFrameworkCore.Design.MigrationsOperations.UpdateDatabase(String targetMigration, String contextType) 
    at Microsoft.EntityFrameworkCore.Tools.Cli.DatabaseUpdateCommand.<>c__DisplayClass0_0.<Configure>b__0() 
    at Microsoft.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args) 
    at Microsoft.EntityFrameworkCore.Tools.Cli.Program.Main(String[] args) 
 The method or operation is not implemented. 

I'm pretty sure this is a provider issue and not an EF core issue, right?

**Comment by how02** [comment]
According to [this](https://github.com/aspnet/EntityFramework/blob/1fa247b038927a7d7438f666dc11253f64e0432d/src/Microsoft.EntityFrameworkCore.Relational/Migrations/MigrationsSqlGenerator.cs#L216), the function throws a NotImplementedException. I guess we have to implement it ourselves [here](https://github.com/SapientGuardian/SapientGuardian.EntityFrameworkCore.MySql/blob/master/src/SapientGuardian.EntityFrameworkCore.MySql/Migrations/MySQLMigrationsSqlGenerator.cs).

I've done quite a lot of migrations during my own usage and never encountered this exception; would you mind providing some code sample so we can reproduce it ?

**Comment by jeffbromberger** [comment]
Hello, thank you for the quick response.

Unfortunately I cannot give out the source to this project, but if you are unable to repro I could try to build a skeleton project which shows this in the next day or so.

Perhaps I can give a few more clues in the meantime.  It seems that the error is related to an alter column operation based on the stack trace.  Then I started looking through my generated migration .cs file.  The property I added is a new property.  I.e. it is an AddColumn operation.  However, I see several AlterColumn calls for fields which did not actually change.

This led me to try another experiment just now.  I wiped all the migrations and started clean.  update-database worked fine.  Then I ran "add-migration mig2" - without changing anything in my project.  Sure enough it generated a migration file with those same AlterColumn statements.  So for whatever reason, it thinks these are always changing.  After seeing this I don't believe it is actually the addition of my new property which causes this particular error, but rather we are getting to the AlterColumn code by this other issue.

Here is the field that is causing the alter columns:

```
    public long RowVersion { get; set; }
```

And I do the following in fluent:

```
        modelBuilder.Entity<Target>()
            .Property(p => p.RowVersion)
            .IsRequired(true)
            .HasColumnType("bigint")
            .IsConcurrencyToken(true)
            .ValueGeneratedNever()
            .HasDefaultValue(1L);
```

In the original migration file I see this:

```
        migrationBuilder.CreateTable(
            name: "TableName",
            columns: table => new
            {
                RowVersion = table.Column<long>(type: "bigint", nullable: false, defaultValue: 1L),
            }
```

and in the second migration I see this:

```
        migrationBuilder.AlterColumn<long>(
            name: "RowVersion",
            table: "TableName",
            type: "bigint",
            nullable: false,
            defaultValue: 1L);
```

So again, I still think the alter column functionality is missing, but I think there is a separate bug that is leading to table alters in cases where nothing has changed.  That could be why you are able to add properties without hitting this issue, because AddColumn may work fine.  Just a guess on my part though!

**Comment by mylemans** [comment]
Just a note: I recently ran in something similar, but I cannot give more feedback other than the fact that another person (me) ran into it. I worked around it back then and I cannot remember all details :/

**Comment by how02** [comment]
The alter column is indeed missing, I'll create a pull request soon to fix that.
I believe that ValueGeneratedNever() somehow make the migration think the column changed, I'm not sure why, I'll have to look deeper into it.

Edit: using both ValueGeneratedNever() and HasDefaultValue(1) at the same time will cause it.

**Comment by SapientGuardian** [comment]
The PR has been merged (thanks @how02!). Please try out version 7.1.14 and let us know how it goes.

**Comment by jeffbromberger** [comment]
Thanks to you both for the quick action on this.

First off, it seems to work now.  Here are the details:

1)  I added a new field to a model, then ran add-migration
2)  In the generated migration cs I saw the new AddColumn as well as the AlterColumns for those fields that didn't really change.
3)  I ran update-database (success)
4)  I went to the server and pulled the query logs.  I only see a query for the addition of the new column.  I do not actually see the 4 alter queries that are shown in the migration file.

So while it still seems like it could be worth looking into the alter issue at some point, it also appears that some other layer in the system is able to determine there isn't actually a change and must be optimizing out those actions.
